arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.h

```diff
@@ -121,12 +203,37 @@ class WeightedTwoSidedEnumerator {
     ProviderType _provider;
 
     PathValidatorType _validator;
-    std::unordered_map<typename Step::VertexType, size_t> _visitedNodes;
+    std::unordered_map<typename Step::VertexType, std::vector<size_t>>
+        _visitedNodes;
     Direction _direction;
     GraphOptions _graphOptions;
   };
   enum BallSearchLocation { LEFT, RIGHT, FINISH };
 
+  /*
+   * A class that is able to store valid shortest paths results.
+   * This is required to be able to check for duplicate paths, as those can be
+   * found during a KShortestPaths graphs search.
+   */
+  class ResultCache {
+   public:
+    ResultCache(Ball& left, Ball& right);
+    ~ResultCache();
+
+    // @brief: returns whether a path could be inserted or not.
+    // True: It will be inserted if this specific path has not been added yet.
+    // False: Could not insert as this path has already been found.
+    auto addResult(CalculatedCandidate const& candidate) -> bool;
+    auto clear() -> void;
+
+   private:
+    Ball& _internalLeft;
+    Ball& _internalRight;
```

Comment:
Where do these balls come from and where do they live?

Reply:
They are equal to "_left" and "_right", only references. As we need access to them to access the provider later.